modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsManagerImpl.java

```diff
@@ -89,86 +157,468 @@ public IgniteStatisticsRepository statisticsRepository() {
-        List<Column> resultList = new ArrayList<>(colNames.length);
+        List<Column> resList = new ArrayList<>(colNames.size());
 
-        for (Column col : columns)
+        for (Column col : cols)
             if (colNamesSet.contains(col.getName()))
-                resultList.add(col);
+                resList.add(col);
 
-        return resultList.toArray(new Column[resultList.size()]);
+        return resList.toArray(new Column[resList.size()]);
     }
 
     /** {@inheritDoc} */
-    @Override public void collectObjectStatistics(String schemaName, String objName, String... colNames)
-            throws IgniteCheckedException {
-        GridH2Table tbl = schemaMgr.dataTable(schemaName, objName);
-        if (tbl == null)
-            throw new IllegalArgumentException(String.format("Can't find table %s.%s", schemaName, objName));
+    @Override public void collectObjectStatistics(
+            String schemaName,
+            String objName,
+            String... colNames
+    ) throws IgniteCheckedException {
+        StatsCollectionFutureAdapter doneFut = new StatsCollectionFutureAdapter();
+        StatsKeyMessage keyMsg = new StatsKeyMessage(schemaName, objName, Arrays.asList(colNames));
+        UUID colId = UUID.randomUUID();
+        StatCollectionStatus status = new StatCollectionStatus(colId, Collections.singletonList(keyMsg),
+            Collections.emptyMap(), doneFut);
+        synchronized (status) {
+            currCollections.updateCollection(colId, s -> status);
+            Map<StatsKeyMessage, int[]> failedPartitions = null;
+            int cnt = 0;
+            do {
+                Collection<StatsCollectionAddrRequest> reqs = currCollections.generateCollectionRequests(colId,
+                    Collections.singletonList(keyMsg), failedPartitions);
+                Map<UUID, StatsCollectionRequest> msgs = reqs.stream().collect(Collectors.toMap(
+                    StatsCollectionAddrRequest::nodeId, StatsCollectionAddrRequest::req));
 
-        if (log.isDebugEnabled())
-            log.debug(String.format("Starting statistics collection by %s.%s object", schemaName, objName));
+                Map<UUID, StatsCollectionRequest> failedMsgs = sendLocalRequests(msgs);
+                failedPartitions = currCollections.extractFailed(failedMsgs.values()
+                    .toArray(new StatsCollectionRequest[0]));
+                if (cnt++ > 10)
+                    throw new IgniteCheckedException(String.format(
+                            "Unable to send all messages to collect statistics by key %s.%s", schemaName, objName));
+            }
+            while (!failedPartitions.isEmpty());
+        }
+        UUID locNode = ctx.discovery().localNode().id();
+        StatsCollectionAddrRequest locReq = status.remainingCollectionReqs().get(locNode);
+        statMgmtPool.submit(() -> processLocal(ctx.discovery().localNode().id(), locReq.req()));
```

Comment:
localNodeId is available via KernalConext.localNodeId().

Reply:
Cool, I'll use it. Fixed